src/highlighter.rs

```diff
@@ -0,0 +1,156 @@
+use crate::colorize::styled_output;
+use crate::colorize::STL;
+
+pub enum PotentialProblem {
+    Head,
+    Tail,
+    Unkn,
+    DistplayOnly,
+}
+
+enum ParseState {
+    BeforeHead,
+    Head,
+    AfterHead,
+
+    Tail,
+    TailLabel,
+    TailDirect,
+    TailIndirect,
+    TailPure,
+
+    AfterTail,
+    Comment,
+}
+
+pub fn code_highlighter(line: usize, source: &str, pp: PotentialProblem) {
+    let mut state = ParseState::BeforeHead;
+    let mut i = 0;
+    let source = source.replace('\t', "    ").replace('\r', "");
+    match pp {
+        PotentialProblem::Head | PotentialProblem::Tail | PotentialProblem::Unkn => {
+            styled_output(&line.to_string(), vec![STL::Red])
+        }
+        PotentialProblem::DistplayOnly => styled_output(&line.to_string(), vec![STL::BrightCyan]),
+    };
+    styled_output("\t", vec![STL::Normal]);
+    styled_output("| ", vec![STL::BrightCyan]);
+
+    while i < source.len() {
+        let c = source.chars().nth(i).unwrap();
+        (state, i) = match state {
```

Comment:
As you've gone so far, didn't you notice an opportunity of introducing an iterator?